apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs

```diff
@@ -537,48 +544,37 @@ private void StartSessionCreationTasksIfNecessary()
-                // we don't worry about it.
 
-                // How many more sessions do we need to create in order to get the pool to the minimum size, after satisfying
-                // pending callers, assuming that all the in-flight session requests succeed, and there are no more requests?
-                int newSessionsToSatisfyPool = (pendingAcquisitionCount + minPoolSize) - (poolSize + inFlightRequests);
-                // How many more sessions *can* we create without going over the maximum number of active sessions?
-                int maxPossibleNewSessions = maxActiveSessions - liveOrRequestedSessionCount;
-
-                int remainingNewSessions = Math.Min(newSessionsToSatisfyPool, maxPossibleNewSessions);
-                while (remainingNewSessions > 0)
-                {
-                    int batchSize = Math.Min(remainingNewSessions, createSessionBatchMaxSize);
-                    Parent.ConsumeBackgroundTask(PrepareNewSessionsAsync(batchSize), "session creation");
-                    remainingNewSessions -= batchSize;
+                    // Determine how many more sessions to create.
+                    // We want to make sure that if all existing and new requests for session creation succeed:
+                    // - All queuing callers will have a session
+                    // - The pool will have at least as many entries as the options specify
+                    // However, we don't want to end up with more than the maximum number of active sessions
+                    // in the options.
+
+                    // In reality, the current pool size should be 0 if pendingAcquisitionCount is non-zero, and
+                    // vice versa. However, we don't strictly enforce this, and the maths works out more simply if
+                    // we don't worry about it.
+
+                    // How many more sessions do we need to create in order to get the pool to the minimum size, after satisfying
+                    // pending callers, assuming that all the in-flight session requests succeed, and there are no more requests?
+                    int newSessionsToSatisfyPool = (_pendingAcquisitions.Count + Options.MinimumPooledSessions) - (poolSize + InFlightSessionCreationCount);
+                    // How many more sessions *can* we create without going over the maximum number of active sessions?
+                    int maxPossibleNewSessions = Options.MaximumActiveSessions - liveOrRequestedSessionCount;
+
+                    int remainingNewSessions = Math.Min(newSessionsToSatisfyPool, maxPossibleNewSessions);
+                    while (remainingNewSessions > 0)
+                    {
+                        int batchSize = Math.Min(remainingNewSessions, Options.CreateSessionMaximumBatchSize);
+                        Parent.ConsumeBackgroundTask(PrepareNewSessionsAsync(batchSize), "session creation");
```

Comment:
I'm slightly nervous now, because we're calling PrepareNewSessionsAsync in the lock, which calls CreatePooledSessionsAsync in the lock... that's a lot of code (in terms of scope) to execute in the lock, and I'm worried that we might end up deadlocking.

It might be okay, but I wonder how feasible it would be to have a loop of "acquire lock, work out how many sessions to create, release lock, call into other code, loop again" until the number of sessions to create is 0? Feel free to argue against this, or to perhaps argue to do it in a separate PR.

Reply:
Okay. Let's defer to another PR. Maybe add a TODO?

(My main cause for concern is that reasoning about how much work happens in the lock is really tricky when it calls async methods, due to the "execute synchronously until the first await that needs to pause" aspect. It's odd to have "this code is sometimes executed in the lock and sometimes not".)

Reply:
Adding a TODO.
(And agree with your concern as well, I'll start a PR straight away to refactor)